feat: add right-click copy context menu to AI chat messages

pulkitbajaj opened the pull request:
### **User description**
Closes #14655
<!-- LINK THE ISSUE WITH THE "Closes" KEYWORD. Example: Closes https://github.com/JabRef/jabref/issues/13109 OR Closes #13109 -->

<!-- In about one to three sentences, describe the changes you have made: what, where, why, ... Summarize changes and DO NOT list modified classes one-by-one. (REPLACE THIS LINE) -->
This PR implements a context menu for the AI chat components, enabling users to copy specific text selections or the entire message via a right-click. Previously, users had no way to interact with the chat text using a mouse.

<!-- NOTE: If your work is not yet complete, please open a draft pull request. In that case, outline your intended next steps. Do you need feedback? Will you continue in parallel? ... -->

### Steps to test
1. Open the AI Chat window.
2. Select a few words of an AI or User message.
3. Right-click the selection and choose "Copy."
4. Paste into a text editor to verify only the selected text was copied.
5. Right-click a message without any selection and choose "Copy."
6. Paste to verify the entire message was copied.

<!-- Describe how reviewers can test this fix/feature. Ideally, think of how you would guide a beginner user of JabRef to try out your change. -->
<!-- You can add screenshots or videos (using Loom - https://www.loom.com or by just adding .mp4 files). -->
<!-- (REPLACE THIS PARAGRAPH) -->

<!-- YOU HAVE TO MODIFY THE ABOVE TEXT TO FIT YOUR PR. OTHERWISE, YOUR PR WILL BE CLOSED WITHOUT FURTHER COMMENT. -->

### Mandatory checks

<!--
Go through the checklist below. It is mandatory, even for a draft pull request.

Keep ALL the items. Replace the dots inside [.] and mark them as follows: 
[x] done 
[ ] TODO (yet to be done)
[/] not applicable
-->

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] I manually tested my changes in running JabRef (always required)
- [ ] I added JUnit tests for changes (if applicable)
- [x] I added screenshots in the PR description (if change is visible to the user)
- [x] I described the change in `CHANGELOG.md` in a way that is understandable for the average user (if change is visible to the user)
- [ ] I checked the [user documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request updating file(s) in <https://github.com/JabRef/user-documentation/tree/main/en>.

<img width="1569" height="354" alt="image" src="https://github.com/user-attachments/assets/f315f927-807f-4376-b86e-e4fa86a03e72" />


___

### **PR Type**
Enhancement


___

### **Description**
- Adds right-click context menu to AI chat messages

- Enables copying selected text or entire message

- Supports UserMessage, AiMessage, and ErrorMessage types

- Integrates with system clipboard for paste functionality


___

### Diagram Walkthrough


```mermaid
flowchart LR
  User["User right-clicks message"]
  Detect["Detect selection via event filter"]
  Menu["Display context menu"]
  Copy["Copy action triggered"]
  Clipboard["Write to system clipboard"]
  User --> Detect
  Detect --> Menu
  Menu --> Copy
  Copy --> Clipboard
```



<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>ChatMessageComponent.java</strong><dd><code>Implement context menu with copy functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

jabgui/src/main/java/org/jabref/gui/ai/components/aichat/chatmessage/ChatMessageComponent.java

<ul><li>Added imports for <code>ContextMenu</code>, <code>MenuItem</code>, <code>Clipboard</code>, and <br><code>ClipboardContent</code><br> <li> Implemented <code>setupContextMenu()</code> method to create context menu with copy <br>functionality<br> <li> Added event filter on <code>markdownContentPane</code> to capture text selection on <br>right-click<br> <li> Supports copying selected text or entire message based on selection <br>state<br> <li> Handles multiple message types (UserMessage, AiMessage, ErrorMessage)</ul>


</details>


  </td>
  <td><a href="https://github.com/JabRef/jabref/pull/14722/files#diff-251343c49554cf5294ddb018a45d71cf53f2c3f8fa34689654462ad9ed493b60">+66/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

calixtus commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/chatmessage/ChatMessageComponent.java:
```diff
@@ -61,6 +69,64 @@ public ChatMessageComponent(ChatMessage chatMessage, Consumer<ChatMessageCompone
+        MenuItem copyItem = new MenuItem(Localization.lang("Copy"));
+
+        copyItem.setOnAction(_ -> {
+            int[] selection = (int[]) contextMenu.getProperties().getOrDefault("selection", new int[] {-1, -1});
```
Isn't there any high level method to get the selected text?

pulkitbajaj replied:
Thank you for the review! I did look for `getSelectedText()` method initially, but I found that `MarkdownTextFlow` does not currently expose one. 

My current implementation uses a `MOUSE_PRESSED` event filter to capture the selection indices because standard handlers often lose the selection state once the `ContextMenu` steals focus in JavaFX. I also added a boundary check (`Math.min`) to handle edge cases where selection indices might exceed the string length, which I found was causing issues when selecting the last word of a message.

I am open for refactoring the underlying `MarkdownTextFlow` to include a `getSelectedText()` method if you believe that is the better architectural approach for JabRef,  but I wanted to provide a robust solution for the AI chat first.

InAnYan commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/chatmessage/ChatMessageComponent.java:
```diff
@@ -61,6 +69,64 @@ public ChatMessageComponent(ChatMessage chatMessage, Consumer<ChatMessageCompone
+                textToCopy = fullText;
+            }
+
+            final Clipboard clipboard = Clipboard.getSystemClipboard();
```
@calixtus @Siedlerchr Is it okay to get a `Clipboard` like this? It's not governed by some `ClipboardService`?

calixtus replied:
we have a ClipBoardManager

pulkitbajaj commented:
I have updated the PR to address the architectural feedback:

- **Logic Separation**: Created `ChatMessageUtils` in the `logic` module to handle content extraction using `Optional<String>`, as requested.
- **Type-Safety**: Refactored `ChatMessageComponent` to use an explicit `selectedText` string property instead of the context menu's metadata map. This makes the selection more intuitive and type-safe.
- **Robustness**: The selection logic extracts the substring during the `MOUSE_PRESSED` phase to ensure the state is captured before focus shifts to the context menu.

Ready for another review!

calixtus commented on jablib/src/main/java/org/jabref/logic/ai/util/ChatMessageUtils.java:
```diff
@@ -0,0 +1,32 @@
+
+        return Optional.ofNullable(content);
+    }
+}
```
I think it is unneccessary to create a new utility class for just that simple method. YAGNI!
Put it into ChatMassageComponent

calixtus replied:
Yeah, could be, maybe. Eventually...
Until then: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

InAnYan commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/chatmessage/ChatMessageComponent.java:
```diff
@@ -37,6 +45,8 @@ public class ChatMessageComponent extends HBox {
 
     private final MarkdownTextFlow markdownTextFlow;
+    private String selectedText = "";
+    private final ClipBoardManager clipBoardManager = new ClipBoardManager(Injector.instantiateModelOrService(StateManager.class));
```
I think it should work, as there is `ViewLoader` in constructor

InAnYan commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/chatmessage/ChatMessageComponent.java:
```diff
@@ -61,6 +73,47 @@ public ChatMessageComponent(ChatMessage chatMessage, Consumer<ChatMessageCompone
+        MenuItem copyItem = new MenuItem(Localization.lang("Copy"));
+
+        copyItem.setOnAction(_ -> {
+            String textToCopy = !this.selectedText.isEmpty()
```
Can you simplify this?

Use `if`, then check if the selected text is empty, then return full message content, otherwise return selected text

InAnYan commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/chatmessage/ChatMessageComponent.java:
```diff
@@ -61,6 +73,47 @@ public ChatMessageComponent(ChatMessage chatMessage, Consumer<ChatMessageCompone
+        contextMenu.getItems().add(copyItem);
+
+        markdownContentPane.addEventFilter(MOUSE_PRESSED, event -> {
+            if (event.isSecondaryButtonDown()) {
```
Can you do early return here?

```java
if (!event.isSecondaryButtonDown()) {
...
```

InAnYan reviewed (changes requested):
Okay, good job, small comments

pulkitbajaj commented:
I have finalized the refactor and verified the logic:

- **Declarative Injection**: Switched to using `@Inject` for the `ClipBoardManager` as requested.
- **Logic Polish**: Implemented the early return pattern in the `MOUSE_PRESSED` event filter to improve readability and simplified the copy action using a standard `if/else` block.

Ready for the review!

InAnYan commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/chatmessage/ChatMessageComponent.java:
```diff
@@ -37,6 +44,9 @@ public class ChatMessageComponent extends HBox {
 
     private final MarkdownTextFlow markdownTextFlow;
+    private String selectedText = "";
+    @Inject
```
Please, follow our style:

```java
private String selectedText = "";

@Inject private ClibBoardManager clipBoardManager;
```

InAnYan commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/chatmessage/ChatMessageComponent.java:
```diff
@@ -61,6 +73,50 @@ public ChatMessageComponent(ChatMessage chatMessage, Consumer<ChatMessageCompone
+        MenuItem copyItem = new MenuItem(Localization.lang("Copy"));
+
+        copyItem.setOnAction(_ -> {
+            String textToCopy;
```
Last thing: can you extract this into a method? So that this method is not too big?

InAnYan reviewed (dismissed):
Okay, almost finished

InAnYan commented:
@pulkitbajaj 
Can you use method `MarkdownTextFlow.copySelectedText`? Or not?

calixtus commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/chatmessage/ChatMessageComponent.java:
```diff
@@ -53,6 +61,8 @@ public ChatMessageComponent() {
         markdownContentPane.getChildren().add(markdownTextFlow);
         markdownContentPane.minHeightProperty().bind(markdownTextFlow.heightProperty());
         markdownContentPane.prefHeightProperty().bind(markdownTextFlow.heightProperty());
+
```
Checkstyle

InAnYan commented on jablib/src/main/java/org/jabref/logic/ai/util/ChatMessageUtils.java:
```diff
@@ -0,0 +1,32 @@
+    private ChatMessageUtils() {
+    }
+
+    public static Optional<String> getContent(ChatMessage chatMessage) {
```
Ah, this is from LangChain4j, we can't change it.

So, I think I can resolve this thread

calixtus replied:
ah i see. thanks for clearance.

subhramit commented:
> I have finalized the refactor and verified the logic:
> 
> * **Selection Stability**: I added a `MOUSE_PRESSED` event filter to consume secondary clicks when a selection is active. In JavaFX, a right-click can trigger a focus change that clears the text highlight before the context menu action executes; consuming the event 'locks' the selection for the copy action.
> * **Logic Delegation**: The 'Copy' action now utilizes `markdownTextFlow.copySelectedText()` as suggested to handle markdown reconstruction.
> * **Method Extraction**: Extracted the fallback copy logic into `copyFullMessage()` to keep `setupContextMenu` concise.

Hi @pulkitbajaj 
Please note that we know how to use LLM and we would not need you a middleman if we wanted to get the issue solved in that manner. Since this PR is complete, we will not be blocking it, but you will not be allowed to contribute further for a certain period of time as a result of violating our [AI usage policy](https://github.com/JabRef/jabref/blob/main/CONTRIBUTING.md#ai-usage-policy).

subhramit reviewed (changes requested):
Please add a Changelog entry and mark the respective mandatory check in the PR description.

jabref-machine commented:
You ticked that you modified `CHANGELOG.md`, but no new entry was found there.

If you made changes that are visible to the user, please add a brief description along with the issue number to the `CHANGELOG.md` file. If you did not, please replace the cross (`[x]`) by a slash (`[/]`) to indicate that no `CHANGELOG.md` entry is necessary. More details can be found in our [Developer Documentation about the changelog](https://devdocs.jabref.org/decisions/0007-human-readable-changelog.html).


<!-- ghprcomment
Check PR CHANGELOG.md
CHANGELOG.md needs to be modified if indicated
-->
